orttraining/orttraining/core/agent/training_agent.cc

```diff
@@ -0,0 +1,136 @@
+﻿// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+#include "orttraining/core/agent/training_agent.h"
+#include "core/session/IOBinding.h"
+#include "orttraining/training_ops/cpu/controlflow/ort_tasks.h"
+
+namespace onnxruntime {
+namespace training {
+
+TrainingAgent::TrainingAgent(InferenceSession *session) : inference_session_(session) {}
+
+TrainingAgent::~TrainingAgent() {
+  // TODO: Properly cancel outstanding background tasks
+  // Following implementation only handle the case where bg_thread is waiting for backward inputs
+  // Background thread can also be in other states, such as running Forward() or running Backward()
+  std::vector<int64_t> run_ids;
+  {
+    std::lock_guard<std::mutex> lock(bg_threads_mutex_);
+    for (auto it = bg_threads_.begin(); it != bg_threads_.end(); ++it) {
+      run_ids.push_back(it->first);
+    }
+  }
+  for (int64_t run_id : run_ids) {
+    if (!onnxruntime::contrib::OrtTasks::GetInstance().TaskIsCompleted(run_id)) {
```

Comment:
>OrtTasks::GetInstance() [](start = 31, length = 23)

why is this singleton necessary? is it possible to put this state in TrainingAgent?

Reply:
oh, i see that OrtTasks is accessed in yield.cc too...
is it possible to not use a singleton (effectively global state)? it kind of hides the interaction between different components and has the problems that global variables have

---
In reply to: [596358823](https://github.com/microsoft/onnxruntime/pull/6973#discussion_r596358823) [](ancestors = 596358823)

Reply:
We also acknowledge that using the singleton is not the best practice for passing data between TrainingAgent and a Kernel. We are working on a new design that will remove the need for OrtTask and running Session::Run() with multiple threads. 

This is the PR in progress: https://github.com/microsoft/onnxruntime/pull/7077